### PR TITLE
Rust dev

### PR DESCRIPTION
**Pull request by dabrain34**

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+fn draw_elements(elements: &Vec<Element>, c: &Context) {
+    for element in elements {
+        c.rectangle(element.position.0, element.position.1, 80.0, 45.0);
+        c.fill().expect("Can not draw into context");
+    }
+}
+
+impl GPSApp {
+    fn new(application: &gtk::Application) -> Result<GPSApp, Box<dyn error::Error>> {
```
Comment (gdesmott) [inline src/app.rs:59]: if you are using `anyhow` you can return `anyhow::Result<>` instead, a bit more ergonomic.

File: src/graph.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    fn default() -> Graph {
+        Graph {
+            elements: vec![],
+            last_x_position: 0,
+        }
+    }
+}
+
+impl Graph {
+    pub fn get_elements(&mut self) -> &Vec<Element> {
```
Comment (gdesmott) [inline src/graph.rs:24]: A more rusty pattern would be:
- name the method `elements()` (we usually don't prefix with `get`).
- maybe return an iterator

File: src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Macro for upgrading a weak reference or returning the given value
+//
+// This works for glib/gtk objects as well as anything else providing an upgrade method
+macro_rules! upgrade_weak {
```
Comment (gdesmott) [inline src/macros.rs:4]: doesn't `glib-rs` already provide such macro?

File: src/pipeline.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
+use gst::prelude::*;
+use gstreamer as gst;
+use std::error;
+
+#[derive(Debug, Ord, Eq, PartialEq, PartialOrd)]
```
Comment (gdesmott) [inline src/pipeline.rs:5]: do you really need `Ord` and `PartialOrd`? `PartialEq` is generally enough.

File: src/pipeline.rs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    SRCTYPE,
+    SINKTYPE,
+}
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct Connection {
+    pad_id: Option<u32>,
+    element_id: Option<u32>,
+}
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct ElementInfo {
```
Comment (gdesmott) [inline src/pipeline.rs:17]: Best to avoid all those `Options` if you can.

File: src/pipeline.rs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            let features = gst::Registry::get_feature_list_by_plugin(&registry, &plugin_name);
+            for feature in features {
+                let mut element = ElementInfo::default();
+                if let Ok(factory) = feature.downcast::<gst::ElementFactory>() {
+                    let feature = factory.upcast::<gst::PluginFeature>();
+
+                    element.name = Some(gst::PluginFeature::get_name(&feature).as_str().to_owned());
+                    element.plugin_name =
+                        Some(gst::Plugin::get_plugin_name(&plugin).as_str().to_owned());
+                    println!(
```
Comment (gdesmott) [inline src/pipeline.rs:76]: for quick debugging you can use something like `dbg!((&element.name, &element.plugin_name));`

File: src/pipeline.rs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    pub fn get_element_description(element_name: &str) -> Result<String, Box<dyn error::Error>> {
+        let mut desc = String::from("");
+        let registry = gst::Registry::get();
+        let feature = gst::Registry::find_feature(
+            &registry,
+            element_name,
+            gst::ElementFactory::static_type(),
+        )
+        .expect("Unable to find the element name");
```
Comment (gdesmott) [inline src/pipeline.rs:96]: you way mant to use `?` is this is meant to be a non-fatal error.

File: src/macros.rs (hunk 3 above)
Comment (dabrain34) [reply src/macros.rs:4]: i dont find it in the doc

Comment (gdesmott) [reply src/macros.rs:4]: Isn't the [clone macro](https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/macro.clone.html) doing something like that? Didn't use it.

File: src/pipeline.rs (hunk 4 above)
Comment (dabrain34) [reply src/pipeline.rs:5]: to perform a elements.sort(), it seems that rustc is complaining if PartialOrd and Ord are not implemented.

Comment (gdesmott) [reply src/pipeline.rs:5]: indeed, you'll need it for sorting.

File: src/macros.rs (hunk 3 above)
Comment (dabrain34) [reply src/macros.rs:4]: not exactly as its working only with glib object as I understand.

File: src/app.rs [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+use crate::graph::{Element, Graph};
+use crate::pipeline::Pipeline;
+use crate::pluginlist;
+
+#[derive(Debug)]
+pub struct GPSAppInner {
+    pub window: gtk::ApplicationWindow,
+    pub drawing_area: DrawingArea,
+    pub builder: Builder,
+    pub pipeline: RefCell<Pipeline>,
```
[the remark this answers is not in the document]
Comment (dabrain34) [reply src/app.rs:21]: Why should I save me from RefCell ?

[the remark this answers is not in the document]
Comment (gdesmott) [reply src/app.rs:21]: Aren't you using `RefCell` for interior mutability (I didn't check)? If you do, protecting the whole inner structure with a Mutex should achieve this as well, I think.

File: src/macros.rs (hunk 3 above)
Comment (dabrain34) [reply src/macros.rs:4]: I keep it for now and then I can remove it later

File: src/app.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dabrain34) [reply src/app.rs:21]: yes I'm using it for interior mutability. What is the tradeoff of using RefCell against Mutex ? Do you have an example ?

[the remark this answers is not in the document]
Comment (gdesmott) [reply src/app.rs:21]: `RefCell` moves borrow checking at runtime, which may lead to panic. But I suppose it's ok if the object isn't shared across threads.

[the remark this answers is not in the document]
Comment (dabrain34) [reply src/app.rs:21]: for now it is not :)

**Pull request closed without merging**